### PR TITLE
Add soft-label support for cross-entropy operator.

### LINKED ISSUE #4080: Add soft-label support for cross-entropy operator.
Label for one instance could be the following three forms:
- [3]: the target class is the 4th.
- [0, 0, 0, 1, 0]: one-hot representation.
- [0.1, 0.2, 0.3, 0.1, 0.3]: soft label representation.

Currently only the first is supported. Will add support for all three forms.

### PR DESCRIPTION
**Pull request by xinghai-sun**
Resolve #4080 
Resolve https://github.com/PaddlePaddle/Paddle/issues/3898

### CODE REVIEW

File: paddle/operators/cross_entropy_op.cc [hunk 1]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
 namespace paddle {
 namespace operators {
 
-class OnehotCrossEntropyOp : public framework::OperatorWithKernel {
+class CrossEntropyOp : public framework::OperatorWithKernel {
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: @luotao1 changes CrossEntropyOp -> OnehotCrossEntropyOp. Please use the new name.

File: paddle/operators/cross_entropy_op.cc [hunk 2]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
+    PADDLE_ENFORCE_EQ(x->dims().size(), 2, "X's rank must be 2.");
+    PADDLE_ASSERT(label->dims().size() == 1 || label->dims().size() == 2);
+    if (label->dims().size() == 2) {
+      // soft cross entropy
+      PADDLE_ENFORCE_EQ(x->dims(), label->dims());
+    } else {
+      // normal cross entropy
+      PADDLE_ENFORCE_EQ(x->dims()[0], label->dims()[0]);
+    }
+    ctx.Output<Tensor>("Y")->Resize({x->dims()[0]});
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]:
```cpp
Output<framework::LoDTensor>
```

Now must use `Output<framework::LoDTensor>` for output in forward and backward `InferShape`.

File: paddle/operators/cross_entropy_op.cc [hunk 3]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
-class OnehotCrossEntropyGradientOp : public framework::OperatorWithKernel {
+class CrossEntropyGradientOp : public framework::OperatorWithKernel {
  public:
   using framework::OperatorWithKernel::OperatorWithKernel;
 
  protected:
   void InferShape(const framework::InferShapeContext &ctx) const override {
-    auto dX = ctx.Output<Tensor>(framework::GradVarName("X"));
-    auto X = ctx.Input<Tensor>("X");
+    auto dx = ctx.Output<Tensor>(framework::GradVarName("X"));
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: `Output< framework::LoDTensor>`

File: paddle/operators/cross_entropy_op.cc [hunk 4]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
+    auto x = ctx.Input<Tensor>("X");
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: Also add not null check for Input(X). Thanks!

File: paddle/operators/cross_entropy_op.cc [hunk 5]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
+    AddInput("Label", "The second input of CrossEntropyOp");
+    AddOutput("Y", "The output of CrossEntropyOp");
     AddComment(R"DOC(
-OnehotCrossEntropy Operator.
+CrossEntropy Operator.
 
-                Y[i] = -log(X[i][j])
+The second input (Label tensor) supports two kinds of shapes:
+1) Rank(Label) = 1, Label[i] indicates the class index for sample i:
+                Y[i] = -log(X[i, Label[i]])
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: Add space before and after formula.

File: paddle/operators/cross_entropy_op.cc [hunk 6]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
 
+2) Rank(Label) = 2, Label[i, j] indicates the soft label of class j
+   for sample i:
+                Y[i] = \sum_j{-Label[i, j] * log(X[i, j])}
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: Add space before and after formula.

File: paddle/operators/cross_entropy_op.cu [hunk 7]
```diff
@@ -21,17 +21,16 @@ namespace operators {
 using Tensor = framework::Tensor;
 
 template <typename T>
-__host__ __device__ T clipping_log(const T x) {
+__host__ __device__ T tolerable_value(const T x) {
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: include `paddle/platform/hostdevice.h`, then use `HOSTDEVICE `.

```cpp
HOSTDEVICE T tolerable_value(const T x) {
```

File: python/paddle/v2/framework/tests/test_cross_entropy_op.py [hunk 8]
```diff
@@ -24,5 +24,26 @@ def test_check_grad(self):
...
+        label /= label.sum(axis=1, keepdims=True)
+        self.inputs = {'X': X, 'Label': label}
+        Y = (-label * numpy.log(X)).sum(axis=1)
+        self.outputs = {'Y': numpy.array(Y).astype("float32")}
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
+        self.check_grad(['X'], 'Y', max_relative_error=0.05)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_cross_entropy_op.py]: Could tune the max_relative_error smaller?

File: paddle/operators/cross_entropy_op.cc (hunk 1 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Since it supports not only one-hot cross-entropy but also soft-label cross-entropy, it would be better to use CrossEntropyOp instead of OnehotCrossEntropyOp.

**Comment by lcy-seso** [comment]
当label变成soft时，已经不再会有离散化的操作，是否可以直接调用 Egien？

File: paddle/operators/cross_entropy_op.cc [hunk 9]
```diff
@@ -17,59 +17,71 @@ limitations under the License. */
...
+class CrossEntropyOp : public framework::OperatorWithKernel {
  public:
   using framework::OperatorWithKernel::OperatorWithKernel;
 
  protected:
   void InferShape(const framework::InferShapeContext &ctx) const override {
-    auto *X = ctx.Input<Tensor>("X");
-    auto *label = ctx.Input<Tensor>("label");
+    auto *x = ctx.Input<Tensor>("X");
+    auto *label = ctx.Input<Tensor>("Label");
```
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cc (hunk 2 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cc (hunk 3 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cc (hunk 4 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cc (hunk 5 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cu (hunk 7 above)
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cu]: Done.

File: python/paddle/v2/framework/tests/test_cross_entropy_op.py (hunk 8 above)
Comment (xinghai-sun) [reply python/paddle/v2/framework/tests/test_cross_entropy_op.py]: Done.

**Comment by xinghai-sun** [comment]
@lcy-seso 我觉得也是可以的，如果愿意容忍同一个if的两个分支：一个走cuda代码，一个走eigen。

**Comment by lcy-seso** [comment]
caff2 分为两为两个分支，可能更关键的是两者在计算上那个更高效。暂时未知。

File: paddle/operators/cross_entropy_op.cc (hunk 2 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cc]: @qingqing01 是否需要按照新的命名规范将输出 Y 改为 Out ？（之前益群在FC中使用Y作为输出时，也被要求改为 Out ，是否需要统一？）

File: paddle/operators/cross_entropy_op.cu [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+#include "paddle/platform/assert.h"
+#include "paddle/platform/hostdevice.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename T>
+HOSTDEVICE T tolerable_value(const T x) {
```
Comment (lcy-seso) [inline paddle/operators/cross_entropy_op.cu]: 我有一个疑问，这个函数已经是 __host__ __device__ 为什么需要在 .cc 中再重复实现一次呢？

File: paddle/operators/cross_entropy_op.cu (hunk 7 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: I have a question here, if this function use `__host__ __device__`  in the declearation, why we need to implement it again in *.cc ?

Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: sorry, please ignore my question ... It's my fault  ...

File: paddle/operators/cross_entropy_op.cu [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  return x;
+}
+
+template <typename T>
+__global__ void CrossEntropyKernel(T* Y, const T* X, const int* label,
+                                   const int N, const int D) {
+  // TOOD(qingqing) define CUDA_1D_KERNEL_LOOP macro in a common file.
+  // CUDA_1D_KERNEL_LOOP(i, N) {
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < N;
+       i += blockDim.x * gridDim.x) {
```
Comment (lcy-seso) [inline paddle/operators/cross_entropy_op.cu]: 我的小疑问请教一下 @qingqing01 
- kernel 函数里面的这个for循环，计算结果不会出错，但逻辑上感觉很怪。
- grid中总会有一些block的thread是多余出来，不会严格对齐到输入数据，这里for循环的作用相当于跳过这些没有对齐的部分。
- `i += blockDim.x * gridDim.x` 循环变量`i`只要增加一次，就会直接超过最大thread数，也就是这个kernel函数实际是并不需要循环多次，只计算输出向量的一个位置，逻辑上等价于判断`i>= batch_size`时，直接return，写成循环有什么好处呢？

Comment (qingqing01) [reply paddle/operators/cross_entropy_op.cu]: 如果针对下面设置grid, threadd的方式，确实不需要for循环。但如果将下面的grid的设置为一个固定的数，就是总共发起固定数目的总线程数，for循环就是有用的，有可能一个线程计算多个输出。这样这个kernel已经处理了边界，就不需要修改了。

```cpp
int block = 512;
int grid = (n + block - 1) / block;
```

Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: 明白啦~ 确实，cross entropy 这个 kernel 比较简单也比较特殊。

File: paddle/operators/cross_entropy_op.cu (hunk 7 above)
Comment (qingqing01) [reply paddle/operators/cross_entropy_op.cu]: 如果换成 HOSTDEVICE，依据HOSTDEVICE的定义：

```cpp
#ifdef __CUDACC__
#define HOSTDEVICE __host__ __device__
#define HOST __host__
#else
#define HOSTDEVICE
#define HOST
#endif
```

确实cpu、gpu可以公用这个`tolerable_value`。

Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cu]: 嗯~ 明白啦~ CPU 下为空。

File: paddle/operators/cross_entropy_op.cu [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+__global__ void SoftCrossEntropyKernel(T* Y, const T* X, const T* label,
+                                       const int N, const int D) {
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < N;
+       i += blockDim.x * gridDim.x) {
+    T sum = static_cast<T>(0);
+    for (int j = 0; j < D; j++) {
+      sum += label[i * D + j] * log(X[i * D + j]);
+    }
+    Y[i] = -tolerable_value(sum);
+  }
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: Put `tolerable_value` aftern `log`:

```cpp
for (int j = 0; j < D; j++) {
  sum += - label[i * D + j] * tolerable_value(log(X[i * D + j]));
}
```

File: paddle/operators/cross_entropy_op.cc [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                      framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The first input of CrossEntropyOp");
+    AddInput("Label", "The second input of CrossEntropyOp");
+    AddOutput("Y", "The output of CrossEntropyOp");
+    AddComment(R"DOC(
+CrossEntropy Operator.
+
+The second input (Label tensor) supports two kinds of shapes:
+1) Rank(Label) = 1, Label[i] indicates the class index for sample i:
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cc]: If modify the int label rank, this doc comments also are needed to modify.

File: paddle/operators/cross_entropy_op.cu [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+template <typename T>
+__global__ void SoftCrossEntropyKernel(T* Y, const T* X, const T* label,
+                                       const int N, const int D) {
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < N;
+       i += blockDim.x * gridDim.x) {
+    T sum = static_cast<T>(0);
+    for (int j = 0; j < D; j++) {
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: Please add `todo` optimization for this kernel.

File: paddle/operators/cross_entropy_op.h [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+    if (label_rank == 2) {
+      // soft cross entropy
+      auto* label_data = ctx.Input<Tensor>("Label")->data<T>();
+      int index = 0;
+      for (int i = 0; i < batch_size; ++i) {
+        T sum = static_cast<T>(0);
+        for (int j = 0; j < class_num; ++j) {
+          sum += label_data[index] * std::log(x_data[index]);
+          y_data[i] = -tolerable_value(sum);
```
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.h]: Use `tolerable_value` before `std::log`.

```cpp
sum += - label_data[index] * tolerable_value(std::log(x_data[index]));
```

File: paddle/operators/cross_entropy_op.cc (hunk 2 above)
Comment (lcy-seso) [reply paddle/operators/cross_entropy_op.cc]: After discussing with @Xreki  , we both prefer "Loss" as the output name rather than "Out".  I think Loss is more meaningful than "Out".

File: paddle/operators/cross_entropy_op.cc [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Label"),
+                            "Input(Label) of CrossEntropyOp must not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Y"),
+                            "Output(Y) of CrossEntropyOp must not be null.");
+
+    auto *x = ctx.Input<Tensor>("X");
+    auto *label = ctx.Input<Tensor>("Label");
+
+    PADDLE_ENFORCE_EQ(x->dims().size(), 2, "X's rank must be 2.");
+    PADDLE_ASSERT(label->dims().size() == 1 || label->dims().size() == 2);
```
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cc]: Done.

File: paddle/operators/cross_entropy_op.cu (hunk 10 above)
[the remark this answers is not in the document]
Comment (xinghai-sun) [reply paddle/operators/cross_entropy_op.cu]: Done.

**Comment by qingqing01** [comment]
Merge this PR, if there is any question, we will fixed later.
